### PR TITLE
Refactor modes into separate modules

### PR DESCRIPTION
**Pull request by ejeschke**
This PR refactors the code for all the modes that exist in the `ginga.Bindings` module into separate modules under `ginga.modes`.  Modes can now be written and understood as having a very similar structure to a plugin--there are explicit constructor, start and stop methods.  Mode docstrings can be written and maintained better to document the modes and the code implementing a mode is much easier to understand since it is encapsulated rather than all mixed together in one huge file.

As part of this refactoring, some potentially useful functions are added to the `ImageView` class, that can be called programmatically for panning by various amounts, rotating by increments, etc.

### CODE REVIEW

**Comment by ejeschke** [comment]
@pllim, do you have time to take a look at this?

**Comment by pllim** [comment]
With `+4,352 −2,572`, I don't even know where to begin. Do you have any suggestion on how to effectively vet this since I am unable to look at all the diff?

**Comment by ejeschke** [comment]
Sorry, @pllim! I realize that maybe this is a lot to digest. It's actually breaking up one large, monolithic module of all the possible user interaction event handling code into smaller, more easily understood modules. *But*--the API is basically the same. `bindings.cfg` files are all backward-compatible. There is a bunch of new documentation added in which I tried to clarify the concept of the modes in the viewer widget, and including some new documentation on developing a mode.

Maybe just start by taking a look at the new documentation and let me know what does not seem clear or needs expanding?

**Comment by pllim** [comment]
Thanks for the clarification! I'll put this in my queue. If this is time sensitive, please give me a deadline.

File: doc/dev_manual/modes.rst [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+===============================
+
+Keyboard and cursor events both have the same callback method signature:
+
+.. code-block:: python
+
+    def kp_handler(self, viewer, event, data_x, data_y)
+    def ms_handler(self, viewer, event, data_x, data_y)
+
+These are class methods, as evidenced by the presence of ``self``.
```
Comment (pllim) [inline doc/dev_manual/modes.rst:165]:
```suggestion
These are instance methods, as evidenced by the presence of ``self``.
```

File: doc/dev_manual/modes.rst [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+* ``event`` : the event which was caught by the trigger
+* ``data_x``, ``data_y`` : the X/Y data coordinates where the cursor was
+  when the event happened (this is also available in the ``event``)
+
+.. note:: The ``data_x`` and ``data_y`` parameters are for backward
+          compatibility.  It is recommended *not* to use them as they
+          may be removed from the callback in a future version.
+          Instead, use the values found in the ``event`` object.
+
+Scroll, pinch and pan events have a slightly different method signature:
```
Comment (pllim) [inline doc/dev_manual/modes.rst:178]:
```suggestion
Scroll, pinch, and pan events have a slightly different method signature:
```

File: doc/dev_manual/modes.rst [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    def sc_handler(self, viewer, event)
+    def pi_handler(self, viewer, event)
+    def pa_handler(self, viewer, event)
+
+These just receive the ``viewer`` and the ``event`` which precipitated the
+callback.
+
+.. note:: To see what attributes are available in each event, see the
+          ``KeyEvent``, ``PointEvent``, ``ScrollEvent``, ``PanEvent`` and
```
Comment (pllim) [inline doc/dev_manual/modes.rst:190]:
```suggestion
          ``KeyEvent``, ``PointEvent``, ``ScrollEvent``, ``PanEvent``, and
```

File: doc/dev_manual/modes.rst [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+          ``PinchEvent`` in the :ref:`api` (look under "ginga.Bindings").
```
Comment (pllim) [inline doc/dev_manual/modes.rst:191]:
```suggestion
          ``PinchEvent`` in the :ref:`api` (look under `ginga.Bindings`).
```

File: doc/dev_manual/modes.rst [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+
```
Comment (pllim) [inline doc/dev_manual/modes.rst:193]:
```suggestion
```

File: doc/manual/modes.rst [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
++++++
+Modes
++++++
+
+As discussed in the Concepts section on :ref:`concepts-modes`,
+modes are a mechanism that allow Ginga to accommodate many key and
+pointer bindings for a large number of possible operations.  
+Modes are associated with the basic Ginga viewer widget and so can be
+used even with the widget standalone in your own programs (i.e. apart
```
Comment (pllim) [inline doc/manual/modes.rst:11]:
```suggestion
used even with the widget standalone in your own programs (i.e., apart
```

**Review by pllim — COMMENTED** [review]
So far, just stylistic comments. I am not done yet but I don't want to lose my work in case of power outage.

File: doc/manual/modes.rst [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+Only one mode (or no mode) can be active at a time in a particular
+viewer widget.  Once a mode is activated, the *mode indicator* should
+turn on in one of the corners of the viewer window to indicate which
+mode is active: 
+
+.. image:: figures/mode_indicator.png
+   :width: 800px
+   :align: center
+
+To exit a mode, simply press the Esc (escape) key.
```
Comment (pllim) [inline doc/manual/modes.rst:34]:
```suggestion
To exit a mode, press the Esc (escape) key.
```

File: doc/manual/modes.rst [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  installed the active canvas.
+
+
+Mode Types
+==========
+
+The mode switching system has a *mode type* which can be set to one of:
+
+* ``oneshot``: The mode is exited by initiating and finishing a cursor drag,
+  or when `Esc` is pressed (if no cursor drag is performed)
```
Comment (pllim) [inline doc/manual/modes.rst:82]:
```suggestion
  or when ``Esc`` is pressed (if no cursor drag is performed).
```

File: doc/manual/modes.rst [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+* ``locked``: The mode is locked until the meta mode key is pressed
+  again (or `Esc`)
```
Comment (pllim) [inline doc/manual/modes.rst:84]:
```suggestion
  again (or ``Esc``).
```

File: doc/manual/modes.rst [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+By default most modes are activated in "locked" type.
+
+.. note::
+
+   When the lock is active it is signified by an additional "[L]"
```
Comment (pllim) [inline doc/manual/modes.rst:90]:
```suggestion
   When the lock is active, it is signified by an additional "[L]"
```

File: doc/manual/modes.rst [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+   (locked) appearing in the mode indicator.  Oneshot mode is
```
Comment (pllim) [inline doc/manual/modes.rst:91]:
```suggestion
   (locked) appearing in the mode indicator; Oneshot mode is
```

File: doc/manual/modes.rst [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+   indicated by "[O]". In the figure above, you can see the mode
+   indicator showing that the viewer is in "contrast" mode, with the
+   mode type as "Lock". In the Reference Viewer, the same information
+   can be seen in the Toolbar plugin, where the lock icon shows
+   the state of the mode type (engaged == 'Lock').
```
Comment (pllim) [inline doc/manual/modes.rst:96]:
```suggestion
   the state of the mode type (``engaged == 'Lock'``).
```

File: doc/manual/modes.rst [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+
+==============
+Standard Modes
+==============
+These are the set of modes that come with Ginga.  Those interested in
+writing their own custom mode should refer to Chapter
```
Comment (pllim) [inline doc/manual/modes.rst:103]:
```suggestion
writing their own custom mode should refer to
```

File: ginga/modes/contrast.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+----------------
+* Esc
+
+Default bindings in mode
+------------------------
+* T : restore contrast to defaults
+* left drag : adjust contrast
+    Interactive shift/stretch colormap (aka contrast and bias).
+    Moving left/right controls shift, up/down controls stretch.
+    Release button when satisfied with the contrast.
```
Comment (pllim) [inline ginga/modes/contrast.py:24]: Line breaks not rendered properly. Try this, maybe?
```suggestion
    * Interactive shift/stretch colormap (aka contrast and bias).
    * Moving left/right controls shift, up/down controls stretch.
    * Release button when satisfied with the contrast.
```

File: doc/quickref.rst [hunk 15]
```diff
@@ -427,6 +404,29 @@ Cmap mode
 | Y                    | Restore the color map to default ('gray').        |
 +----------------------+---------------------------------------------------+
 
+.. _naxis_mode:
+
+Naxis mode
+==========
+
+.. note:: Naxis mode is used when viewing data that has more than 2
+          dimensions (e.g. data cubes).
```
Comment (pllim) [inline doc/quickref.rst:413]:
```suggestion
          dimensions (e.g., data cubes).
```

**Review by pllim — COMMENTED** [review]
More stylistic comments.

I am unable to go through all the LOC touched here, so I only focused on the user docs, as you suggested. If this is working for your users, feel free to merge. Thanks!

**Comment by ejeschke** [comment]
> I am unable to go through all the LOC touched here, so I only focused on the user docs, as you suggested. If this is working for your users, feel free to merge.

Thanks for the review, @pllim!

**Pull request merged**